src/umpire/strategy/DynamicPool.cpp

```diff
@@ -13,382 +13,341 @@
-    newBlock->blockSize = 0;
-    newBlock->next = curr->next;
-    next = newBlock;
-    curr->size = size;
-  }
+void* DynamicPool::allocate(std::size_t bytes)
+{
+  UMPIRE_LOG(Debug, "(bytes=" << bytes << ")");
 
-  if (prev) prev->next = next;
-  else freeBlocks = next;
-}
+  const std::size_t actual_bytes = round_up(bytes, m_align_bytes);
+  Pointer ptr{nullptr};
 
-void DynamicPool::releaseBlock(struct Block *curr, struct Block *prev) {
-  assert(curr != nullptr);
+  // Check if the previous block is a match
+  SizeMap::const_iterator iter{findFreeBlock(actual_bytes)};
 
-  UMPIRE_LOG(Debug, "(" << curr << ", " << prev << ")");
+  // This is optional, but it might help the growth of the pool...
+  if (iter == m_free_map.end()) {
+    doCoalesce();
+    iter = findFreeBlock(actual_bytes);
+  }
 
-  if (prev) prev->next = curr->next;
-  else usedBlocks = curr->next;
+  if (iter != m_free_map.end()) {
+    // Found this acceptable address pair
+    bool is_head;
+    std::tie(ptr, is_head) = iter->second;
 
-  // Find location to put this block in the freeBlocks list
-  prev = nullptr;
-  for ( struct Block *temp = freeBlocks ; temp && temp->data < curr->data ; temp = temp->next )
-    prev = temp;
+    // Add used map
+    insertUsed(ptr, actual_bytes, is_head);
 
-  // Keep track of the successor
-  struct Block *next = prev ? prev->next : freeBlocks;
+    // Remove the entry from the free map
+    const std::size_t free_size = iter->first;
+    m_free_map.erase(iter);
 
-  // Check if prev and curr can be merged
-  if ( prev && prev->data + prev->size == curr->data && !curr->blockSize ) {
-    prev->size = prev->size + curr->size;
-    blockPool.deallocate(curr); // keep data
-    curr = prev;
-  }
-  else if (prev) {
-    prev->next = curr;
-  }
-  else {
-    freeBlocks = curr;
-  }
+    m_curr_bytes += actual_bytes;
 
-  // Check if curr and next can be merged
-  if ( next && curr->data + curr->size == next->data && !next->blockSize ) {
-    curr->size = curr->size + next->size;
-    curr->next = next->next;
-    blockPool.deallocate(next); // keep data
-  }
-  else {
-    curr->next = next;
-  }
-}
-
-std::size_t DynamicPool::freeReleasedBlocks() {
-  UMPIRE_LOG(Debug, "()");
-  // Release the unused blocks
-  struct Block *curr = freeBlocks;
-  struct Block *prev = nullptr;
-
-  std::size_t freed = 0;
-
-  while ( curr ) {
-    struct Block *next = curr->next;
-    // The free block list may contain partially released released blocks.
-    // Make sure to only free blocks that are completely released.
-    //
-    if ( curr->size == curr->blockSize ) {
-      totalBlocks -= 1;
-      totalBytes -= curr->blockSize;
-      freed += curr->blockSize;
-      m_allocator->deallocate(curr->data);
-
-      if ( prev )   prev->next = curr->next;
-      else          freeBlocks = curr->next;
-
-      blockPool.deallocate(curr);
+    const int64_t left_bytes{static_cast<int64_t>(free_size - actual_bytes)};
+    if (left_bytes > m_align_bytes) {
```

Comment:
Just marking this with a comment so that we can discuss the block's lifecycle during our meeting today.